### PR TITLE
feat: better error handling

### PR DESCRIPTION
**Pull request by AtifChy**
previously it was very tedious to handle errors, i simplified the process by using `thiserror` crate.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -45,9 +45,23 @@ use std::fmt::Write;
...
 use std::str::FromStr;
 
-pub use windows::core::{Error, Result, HSTRING};
+pub use windows::core::HSTRING;
 pub use windows::UI::Notifications::ToastNotification;
 
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum ToastError {
```
Comment (amrbashir) [inline src/lib.rs]: I think we can just call it `Error`

File: src/lib.rs [hunk 2]
```diff
@@ -447,12 +461,13 @@ impl Toast {
 
     // HACK: f is static so that we know the function is valid to call.
     //       this would be nice to remove at some point
-    pub fn on_activated<F: FnMut(Option<String>) -> Result<()> + Send + 'static>(
-        mut self,
-        mut f: F,
-    ) -> Self {
+    pub fn on_activated<F>(mut self, mut f: F) -> Self
```
Comment (amrbashir) [inline src/lib.rs:462]: any reason to change the closure signature?

File: src/lib.rs [hunk 3]
```diff
@@ -564,11 +579,11 @@ impl Toast {
         )))?;
 
         // Create the toast
-        ToastNotification::CreateToastNotification(&toast_xml)
+        Ok(ToastNotification::CreateToastNotification(&toast_xml)?)
```
Comment (amrbashir) [inline src/lib.rs]:
```suggestion
        ToastNotification::CreateToastNotification(&toast_xml).map_err(Into::into)
```

File: src/lib.rs [hunk 4]
```diff
@@ -582,7 +597,7 @@ impl Toast {
             ToastNotificationManager::CreateToastNotifierWithId(&HSTRING::from(&self.app_id))?;
 
         // Show the toast.
-        let result = toast_notifier.Show(&toast_template);
+        let result = Ok(toast_notifier.Show(&toast_template)?);
```
Comment (amrbashir) [inline src/lib.rs]:
```suggestion
        let result = toast_notifier.Show(&toast_template).map_err(Into::into);
```

File: src/lib.rs [hunk 5]
```diff
@@ -489,15 +504,15 @@ impl Toast {
     ///         Some(ToastDismissalReason::TimedOut) => println!("TimedOut"),
     ///         _ => println!("Unknown"),
     ///     }
-    ///     Ok(())
     /// }).show().expect("notification failed");
     /// ```
-    pub fn on_dismissed<F: Fn(Option<ToastDismissalReason>) -> Result<()> + Send + 'static>(
+    pub fn on_dismissed<F: Fn(Option<ToastDismissalReason>) + Send + 'static>(
```
Comment (amrbashir) [inline src/lib.rs]: any reason to change the closure signature?

File: Cargo.toml [hunk 6]
```diff
@@ -19,6 +19,7 @@ default-target = "x86_64-pc-windows-msvc"
 
 [dependencies]
 quick-xml = "0.31"
+thiserror = "1.0.61"
```
Comment (amrbashir) [inline Cargo.toml]:
```suggestion
thiserror = "1.0"
```

**Review by amrbashir — CHANGES_REQUESTED** [review]
Thanks for your contributions, left a few questions and nitpicks

File: src/lib.rs (hunk 1 above)
Comment (AtifChy) [reply src/lib.rs]: okey

File: src/lib.rs (hunk 2 above)
Comment (AtifChy) [reply src/lib.rs:462]: while i was testing it, the compiler was giving error about windows::core::Result not being found, so i removed it and it was working fine. But i reversed this change in recent commit.

**Review by AtifChy — COMMENTED** [review]
I made all the necessary changes

File: src/lib.rs [hunk 7]
```diff
@@ -45,9 +45,23 @@ use std::fmt::Write;
...
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Windows API error: {0}")]
+    Os(#[from] windows::core::Error),
+    #[error("IO error: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("Unknown error")]
+    Unknown,
```
Comment (amrbashir) [inline src/lib.rs]:
```suggestion
```

**Review by amrbashir — APPROVED** [review]

**Pull request merged**